PermissionsActivity: java.lang.IllegalArgumentException: `PERMISSIONS_SET_EXTRA` not set

https://ankidroid.org/acra/app/1/bug/399530/report/42454fd6-c30a-41fa-821f-cbfba94762f3

```
java.lang.RuntimeException: Unable to start activity ComponentInfo{com.ichi2.anki/com.ichi2.anki.ui.windows.permissions.PermissionsActivity}: java.lang.IllegalArgumentException: PERMISSIONS_SET_EXTRA not set
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3645)
	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3782)
	at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:101)
	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2307)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loopOnce(Looper.java:201)
	at android.os.Looper.loop(Looper.java:288)
	at android.app.ActivityThread.main(ActivityThread.java:7872)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)
Caused by: java.lang.IllegalArgumentException: PERMISSIONS_SET_EXTRA not set
	at com.ichi2.anki.ui.windows.permissions.PermissionsActivity.onCreate(PermissionsActivity.kt:64)
	at android.app.Activity.performCreate(Activity.java:8305)
	at android.app.Activity.performCreate(Activity.java:8284)
	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1417)
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3626)
	... 12 more
```

--- a/AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/permissions/PermissionsActivity.kt
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/permissions/PermissionsActivity.kt
@@ -26,12 +26,14 @@ import com.ichi2.anki.AnkiActivity
 import com.ichi2.anki.PermissionSet
 import com.ichi2.anki.R
 import com.ichi2.anki.databinding.ActivityPermissionsBinding
+import com.ichi2.anki.showThemedToast
 import com.ichi2.anki.ui.windows.permissions.PermissionsFragment.Companion.HAS_ALL_PERMISSIONS_KEY
 import com.ichi2.anki.ui.windows.permissions.PermissionsFragment.Companion.PERMISSIONS_FRAGMENT_RESULT_KEY
 import com.ichi2.anki.utils.ext.setFragmentResultListener
 import com.ichi2.themes.Themes
 import com.ichi2.themes.setTransparentStatusBar
 import dev.androidbroadcast.vbpd.viewBinding
+import timber.log.Timber
 
 /**
  * Screen responsible for getting permissions from the user.
@@ -60,10 +62,15 @@ class PermissionsActivity : AnkiActivity(R.layout.activity_permissions) {
 
         binding.continueButton.setOnClickListener { finish() }
 
-        val permissionSet =
-            requireNotNull(IntentCompat.getParcelableExtra(intent, PERMISSIONS_SET_EXTRA, PermissionSet::class.java)) {
-                "PERMISSIONS_SET_EXTRA not set"
-            }
+        // #20881: Activity should not be launchd without extras
+        val permissionSet = IntentCompat.getParcelableExtra(intent, PERMISSIONS_SET_EXTRA, PermissionSet::class.java)
+        if (permissionSet == null) {
+            Timber.w("PERMISSIONS_SET_EXTRA not set; finishing")
+            showThemedToast(this, R.string.something_wrong, false)
+            setResult(RESULT_CANCELED)
+            finish()
+            return
+        }
         val permissionsFragment =
             requireNotNull(permissionSet.permissionsFragment?.getDeclaredConstructor()?.newInstance()) {
                 "invalid permissionsFragment"

--- a/AnkiDroid/src/test/java/com/ichi2/anki/ui/windows/permissions/PermissionsActivityTest.kt
+++ b/AnkiDroid/src/test/java/com/ichi2/anki/ui/windows/permissions/PermissionsActivityTest.kt
@@ -15,6 +15,8 @@
  */
 package com.ichi2.anki.ui.windows.permissions
 
+import android.content.Context
+import android.content.Intent
 import androidx.appcompat.widget.AppCompatButton
 import androidx.fragment.app.commitNow
 import androidx.test.core.app.ActivityScenario
@@ -26,14 +28,18 @@ import com.ichi2.anki.R
 import com.ichi2.anki.RobolectricTest
 import com.ichi2.testutils.HamcrestUtils.containsInAnyOrder
 import org.hamcrest.MatcherAssert.assertThat
+import org.hamcrest.Matchers.equalTo
 import org.junit.Test
 import org.junit.runner.RunWith
+import org.robolectric.shadows.ShadowToast
+import kotlin.test.assertEquals
+import kotlin.test.assertFailsWith
 
 @RunWith(AndroidJUnit4::class)
 class PermissionsActivityTest : RobolectricTest() {
     @Test
     fun testActivityCantBeClosedByBackButton() {
-        testActivity(ARBITRARY_PERMISSION_SET) { activity ->
+        testActivity { activity ->
             activity.onBackPressedDispatcher.onBackPressed()
             assertThat("activity is not finishing", !activity.isFinishing)
         }
@@ -41,7 +47,7 @@ class PermissionsActivityTest : RobolectricTest() {
 
     @Test
     fun testOnClickingContinueActivityFinishes() {
-        testActivity(ARBITRARY_PERMISSION_SET) { activity ->
+        testActivity { activity ->
             activity.setContinueButtonEnabled(true)
             activity.findViewById<AppCompatButton>(R.id.continue_button).performClick()
             assertThat("activity is finishing", activity.isFinishing)
@@ -49,8 +55,17 @@ class PermissionsActivityTest : RobolectricTest() {
     }
 
     @Test
+    fun `error toast is shown if PERMISSIONS_SET_EXTRA is missing`() {
+        testInvalidActivityFinishes()
+        assertThat(
+            ShadowToast.getTextOfLatestToast(),
+            equalTo(getResourceString(R.string.something_wrong)),
+        )
+    }
+
+    @Test
     fun `Each screen starts normally and has the same permissions of a PermissionSet`() {
-        testActivity(ARBITRARY_PERMISSION_SET) { activity ->
+        testActivity { activity ->
             for (permissionSet in PermissionSet.entries) {
                 val fragment = permissionSet.permissionsFragment?.getDeclaredConstructor()?.newInstance() ?: continue
                 activity.supportFragmentManager.commitNow {
@@ -63,15 +78,22 @@ class PermissionsActivityTest : RobolectricTest() {
         }
     }
 
+    private fun testInvalidActivityFinishes() {
+        val ex = assertFailsWith<NullPointerException> { testActivity(permissionSet = null) { } }
+        assertEquals("Cannot run onActivity since Activity has been destroyed already", ex.message)
+    }
+
     private fun testActivity(
-        permissionSet: PermissionSet,
+        permissionSet: PermissionSet? = ARBITRARY_PERMISSION_SET,
         action: ActivityAction<PermissionsActivity>,
     ) {
+        val context = ApplicationProvider.getApplicationContext<Context>()
         val intent =
-            PermissionsActivity.getIntent(
-                ApplicationProvider.getApplicationContext(),
-                permissionSet,
-            )
+            if (permissionSet != null) {
+                PermissionsActivity.getIntent(context, permissionSet)
+            } else {
+                Intent(context, PermissionsActivity::class.java)
+            }
         ActivityScenario.launch<PermissionsActivity>(intent).use { scenario ->
             scenario.onActivity { activity ->
                 action.perform(activity)

